### PR TITLE
CODENVY-581: Add Audit IM command

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
Add `Audit` Installation manager command

### What issues does this PR fix or reference?
https://github.com/codenvy/codenvy/issues/581

### Tests written?
Yes

@dmytro-ndp please review

### CODE REVIEW

File: cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+fetchJsonParameter "id"
+USER_ID=${OUTPUT}
+
+authWithoutRealmAndServerDns "user1" "pwd123ABC"
+# create workspace 1
+WORKSPACE_NAME="workspace-1"
+doPost "application/json" "{\"defaultEnv\":\"default\",\"commands\":[{\"commandLine\":\"mvn clean install -f $\{current.project.path}\",\"name\":\"build\",\"type\":\"mvn\",\"attributes\":{}}],\"projects\":[],\"name\":\"${WORKSPACE_NAME}\",\"environments\":{\"default\":{\"recipe\":{\"location\":\"codenvy/ubuntu_jdk8\",\"type\":\"dockerimage\"},\"machines\":{\"dev-machine\":{\"servers\":{},\"agents\":[\"org.eclipse.che.terminal\",\"org.eclipse.che.ws-agent\",\"org.eclipse.che.ssh\"],\"attributes\":{\"memoryLimitBytes\":1610612736},\"source\":{\"type\":\"dockerfile\",\"content\":\"FROM codenvy/ubuntu_jdk8\"}}}}},\"links\":[],\"description\":null}" "http://${HOST_URL}/api/workspace/?token=${TOKEN}"
+
+# create workspace 2
+WORKSPACE_NAME="workspace-2"
```
Comment (dmytro-ndp) [inline cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh]: From the readability point it's better to use separate constants for each workspace name and user name: WORKSPACE_NAME_1,  WORKSPACE_NAME_2,...; USER_1, USER_2, ....

File: cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+
+# try to execute audit from user
+executeIMCommand "--valid-exit-code=1" "audit"
+validateExpectedString ".*Please,.login.into.Codenvy.*"
+
+#execute audit from admin
+executeIMCommand "login" "admin" "password"
+executeIMCommand "audit"
+
+validateExpectedString ".*Number.of.all.users:.3.*Number.of.users.licensed:.10.*Date.when.license.expires:.31.December.2050.*admin@codenvy.onprem.is.owner.of.0.workspaces.and.has.permissions.in.0.workspaces.*user1.im.test@email.com.is.owner.of.2.workspaces.and.has.permissions.in.2.workspaces.*workspace-1,.is.owner:.true,.permissions:.\[read,.use,.run,.configure,.setPermissions,.delete\].*workspace-2,.is.owner:.true,.permissions:.\[read,.use,.run,.configure,.setPermissions,.delete\].*user2.im.test@email.com.is.owner.of.1.workspace.and.has.permissions.in.2.workspaces.*workspace-2,.is.owner:.false,.permissions:.\[read,.use,.run,.configure\].*workspace-1,.is.owner:.true,.permissions:.\[read,.use,.run,.configure,.setPermissions,.delete\]"
```
Comment (dmytro-ndp) [inline cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh]: From the changeability point it's better to use constants in stead of hardcoded names **workspace-1**, **user2** and so on.

File: cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    protected void doExecuteCommand() throws IOException {
+        try {
+            getFacade().requestAuditReport(getCodenvyOnpremPreferences().getAuthToken(), getCodenvyOnpremPreferences().getUrl());
+        } catch (PreferenceNotFoundException e) {
+            getConsole().printErrorAndExit("Please, login into Codenvy");
+            return;
+        }
+
+        File auditDirectory =
```
Comment (dmytro-ndp) [inline cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java]: From the performance and readability points int's better to extract `File auditDtirectory` field with initialization in constructor.

File: cdec/installation-manager-core/src/main/java/com/codenvy/im/facade/InstallationManagerFacade.java [hunk 4]
```diff
@@ -105,6 +115,21 @@ public InstallationManagerFacade(HttpTransport transport,
...
+     * Makes a request to download Audit report and save it to audit directory.
+     */
+    public void requestAuditReport(String authToken, String host) throws IOException {
+
+        Path path = Paths.get(InjectorBootstrap.INJECTOR.getInstance(Key.get(String.class, Names.named("installation-manager.audit_dir"))));
+
+        if (configManager.getHostUrl().equals(host)) {
+            transport.downloadWithoutProxy(host + "/api/audit", authToken, path, MediaType.TEXT_PLAIN);
+        } else {
+            transport.download(host + "/api/audit?token=" + authToken, path, MediaType.TEXT_PLAIN);
```
Comment (dmytro-ndp) [inline cdec/installation-manager-core/src/main/java/com/codenvy/im/facade/InstallationManagerFacade.java]: Method 
```
com.codenvy.im.utils.HttpTransport#download(String path,
                          String method,
                          String accessToken,
                          String expectedContentType,
                          Path destinationDir,
                          boolean noProxy))
```
is more suitable here when we use **accessToken**.

**Comment by dmytro-ndp** [comment]
@vinokurig: **_TestConfigManager_** failed with an error:
```
Failed tests: 
com.codenvy.im.managers.TestConfigManager.setUp(com.codenvy.im.managers.TestConfigManager)
  Run 1: TestConfigManager.setUp:72 UnfinishedStubbing 
Unfinished stubbing detected he...
  Run 2: PASS
  Run 3: PASS
  Run 4: PASS
  Run 5: PASS
  Run 6: PASS
  Run 7: PASS
  Run 8: PASS
  Run 9: PASS
  Run 10: PASS
  Run 11: PASS
  Run 12: PASS
  Run 13: PASS
  Run 14: PASS
  Run 15: PASS
  Run 16: PASS

  TestConfigManager.testLoadDefaultCdecConfigLoadError » Test 
Expected exceptio...
```

**Comment by dmytro-ndp** [comment]
@tolusha: please, take a look at this request as well.

File: cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+printAndLog "TEST CASE: Audit"
+vagrantUp ${SINGLE_NODE_VAGRANT_FILE}
+
+# install Codenvy on-prem
+installCodenvy ${LATEST_CODENVY_VERSION}
+validateInstalledCodenvyVersion ${LATEST_CODENVY_VERSION}
+
+# install IM CLI
+installImCliClient ${LATEST_IM_CLI_CLIENT_VERSION}
```
Comment (vinokurig) [inline cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh]: fixed

File: cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh (hunk 1 above)
Comment (vinokurig) [inline cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh]: done

File: cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh (hunk 2 above)
Comment (vinokurig) [inline cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh]: done

File: cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java (hunk 3 above)
Comment (vinokurig) [inline cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java]: done

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java [hunk 6]
```diff
@@ -109,7 +110,8 @@ public void testLoadDefaultSingleServerCdecConfig() throws Exception {
                                              "b=2\n" +
                                              "c=\n" +
                                              "d=3\\n4\n");
-        doReturn(properties).when(transport).download(endsWith("codenvy-single-server-properties/3.1.0"), any(Path.class));
+        doReturn(properties).when(transport).download(endsWith("codenvy-single-server-properties/3.1.0"), any(Path.class),
+                                                      MediaType.APPLICATION_OCTET_STREAM);
```
Comment (dmytro-ndp) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: Leads to _Mockito_ error `Unfinished stubbing detected`.
Fixup: `eq(MediaType.APPLICATION_OCTET_STREAM)`

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java [hunk 7]
```diff
@@ -131,7 +133,8 @@ public void testLoadDefaultMultiServerCdecConfig() throws Exception {
                                              "user_ldap_dn=dc=codenvy-enterprise,dc=com\n" +
                                              "billing_resources_refill_cron=0 15 0 * * ?\n" +
                                              "admin_ldap_dn=dc=codenvycorp,dc=com\n");
-        doReturn(properties).when(transport).download(endsWith("codenvy-multi-server-properties/3.1.0"), any(Path.class));
+        doReturn(properties).when(transport).download(endsWith("codenvy-multi-server-properties/3.1.0"), any(Path.class),
+                                                      MediaType.APPLICATION_OCTET_STREAM);
```
Comment (dmytro-ndp) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: Leads to _Mockito_ error `Unfinished stubbing detected`.
Fixup: `eq(MediaType.APPLICATION_OCTET_STREAM)`

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java [hunk 8]
```diff
@@ -146,7 +149,8 @@ public void testLoadDefaultMultiServerCdecConfig() throws Exception {
     @Test(expectedExceptions = IOException.class,
             expectedExceptionsMessageRegExp = "Can't download installation properties. error")
     public void testLoadDefaultCdecConfigTransportError() throws Exception {
-        doThrow(new IOException("error")).when(transport).download(endsWith("codenvy-multi-server-properties/3.1.0"), any(Path.class));
+        doThrow(new IOException("error")).when(transport).download(endsWith("codenvy-multi-server-properties/3.1.0"), any(Path.class),
+                                                                   MediaType.APPLICATION_OCTET_STREAM);
```
Comment (dmytro-ndp) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: Leads to _Mockito_ error `Unfinished stubbing detected`.
Fixup: `eq(MediaType.APPLICATION_OCTET_STREAM)`

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java [hunk 9]
```diff
@@ -156,7 +160,8 @@ public void testLoadDefaultCdecConfigLoadError() throws Exception {
         Path properties = Paths.get("target/test.properties");
         FileUtils.write(properties.toFile(), "a=1\n" +
                                              "b=2\n");
-        doReturn(properties).when(transport).download(endsWith("codenvy-multi-server-properties/3.1.0"), any(Path.class));
+        doReturn(properties).when(transport).download(endsWith("codenvy-multi-server-properties/3.1.0"), any(Path.class),
+                                                      MediaType.APPLICATION_OCTET_STREAM);
```
Comment (dmytro-ndp) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: Leads to _Mockito_ error `Unfinished stubbing detected`.
Fixup: `eq(MediaType.APPLICATION_OCTET_STREAM)`

File: cdec/installation-manager-core/src/main/java/com/codenvy/im/facade/InstallationManagerFacade.java (hunk 4 above)
Comment (vinokurig) [inline cdec/installation-manager-core/src/main/java/com/codenvy/im/facade/InstallationManagerFacade.java]: done

File: cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+# try to execute audit from user
+executeIMCommand "--valid-exit-code=1" "audit"
+validateExpectedString ".*Please,.login.into.Codenvy.*"
+
+#execute audit from admin
+executeIMCommand "login" "admin" "password"
+executeIMCommand "audit"
+
+validateExpectedString ".*Number.of.all.users:.3.*Number.of.users.licensed:.10.*Date.when.license.expires:.31.December.2050.*admin@codenvy.onprem.is.owner.of.0.workspaces.and.has.permissions.in.0.workspaces.*user1.im.test@email.com.is.owner.of.2.workspaces.and.has.permissions.in.2.workspaces.*${USER_1_WORKSPACE_1_NAME},.is.owner:.true,.permissions:.\[read,.use,.run,.configure,.setPermissions,.delete\].*${USER_1_WORKSPACE_2_NAME},.is.owner:.true,.permissions:.\[read,.use,.run,.configure,.setPermissions,.delete\].*user2.im.test@email.com.is.owner.of.1.workspace.and.has.permissions.in.2.workspaces.*${USER_1_WORKSPACE_2_NAME},.is.owner:.false,.permissions:.\[read,.use,.run,.configure\].*${USER_2_WORKSPACE_1_NAME},.is.owner:.true,.permissions:.\[read,.use,.run,.configure,.setPermissions,.delete\]"
```
Comment (vinokurig) [inline cdec/installation-manager-tests/src/main/resources/bin/codenvy/audit/test-audit.sh]: done

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java (hunk 6 above)
Comment (vinokurig) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: done

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java (hunk 7 above)
Comment (vinokurig) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: done

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java (hunk 8 above)
Comment (vinokurig) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: done

File: cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java (hunk 9 above)
Comment (vinokurig) [inline cdec/installation-manager-core/src/test/java/com/codenvy/im/managers/TestConfigManager.java]: done

File: cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+import java.io.IOException;
+import java.nio.charset.Charset;
+import java.util.List;
+
+/**
+ * Installation manager Audit command.
+ *
+ * @author Igor Vinokur
+ */
+@Command(scope = "codenvy", name = "audit", description = "Download Audit report and print it to screen")
```
Comment (tolusha) [inline cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java]: on the screen

File: cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+            return;
+        }
+
+        File[] reports = auditDirectory.listFiles();
+        if (reports == null || reports.length == 0) {
+            getConsole().printErrorAndExit("Audit directory is empty");
+            return;
+        }
+
+        File lastModifiedFile = reports[0];
```
Comment (tolusha) [inline cdec/installation-manager-cli-bundle/src/main/java/com/codenvy/im/cli/command/AuditCommand.java]: Use ```Stream.of(T ... values)``` to create stream and to sort and to find needed report file.

**Pull request merged**